Batch Requests

ahageali opened the pull request:
While experimental, I have tested this through the use of urlshortener apis and drive apis and this does appear to work for normal GET and POST requests. I have also added some documentation to provide a simple way to call the batch api. As a result however, the api requests no longer actually make their requests when they are called without a callback. 

I have also _tried_ to keep as similar a style as possible to all the pre-existing code so I apologize if I made any horrendous mistakes. Let me know if you want any changes!

ryanseys commented on lib/googleapis.js:
```diff
@@ -65,6 +67,80 @@ GoogleApis.prototype.addAPIs = function(apis) {
+    //if there is a query string, append it to the end of the path
+    if (request.useQuerystring && request.qs) {
+      var firstTime = true;
+      for (var key in request.qs) {
```
you could probably use a module like `qs` to handle this for you.

ryanseys commented on lib/googleapis.js:
```diff
@@ -65,6 +67,80 @@ GoogleApis.prototype.addAPIs = function(apis) {
+GoogleApis.prototype.batch = function(params, callback) {
+  var multipart = [],
+      rootUrl = 'https://www.googleapis.com';
+      //rootUrl = 'http://localhost:8080';
```
Remove

ryanseys commented on lib/googleapis.js:
```diff
@@ -65,6 +67,80 @@ GoogleApis.prototype.addAPIs = function(apis) {
+  }
+  
+  reqs.forEach(function(request, pos) {
+    //TODO check if vital components of a request exist
```
Don't leave TODOs in here. Not sure what this means.

ryanseys commented on lib/googleapis.js:
```diff
@@ -65,6 +67,80 @@ GoogleApis.prototype.addAPIs = function(apis) {
+
+  return new BatchTransporter(authClient ? authClient : new DefaultTransporter()).request({
+    method: 'POST',
+    uri: rootUrl + '/batch',
```
So this is just https://www.googleapis.com/batch ?

ryanseys commented on lib/apirequest.js:
```diff
@@ -85,10 +85,11 @@ function createAPIRequest(parameters, callback) {
   // If the params are not present, and callback was passed instead,
   // use params as the callback and create empty params.
   if (typeof params === 'function') {
+    missingCallback = false;
```
No need for this if you have `missingCallback = !callback;` later.

ryanseys commented on lib/apirequest.js:
```diff
@@ -184,6 +186,10 @@ function createAPIRequest(parameters, callback) {
   delete options.auth; // is overridden by our auth code
   delete options.params; // We handle params ourselves and Request does not recognise 'params'
 
+  if (missingCallback) {
```
Just use !callback I'd say.

ahageali replied:
Was originally going to do that but the createCallback function guarantees that !callback will always evaluate to false

ryanseys replied:
You're right. If no callback is given then it's no longer an error so we can get rid of createCallback and logError function in this file.

ryanseys commented on lib/apirequest.js:
```diff
@@ -184,6 +186,10 @@ function createAPIRequest(parameters, callback) {
   delete options.params; // We handle params ourselves and Request does not recognise 'params'
 
+  if (missingCallback) {
+    return options;
```
Will you also need to return the `body` potentially?

ahageali replied:
Most likely but I'll need to find an API i can play with that uses body before figuring out how to incorporate it

ahageali replied:
Regarding this, I've found that most scenarios that require the `body` variable causes the request to be a multipart request. 
For example, `drive.insert({media: {body: readStream}, uploadType: 'multipart', resource: {title: 'Test'}})` will cause a multipart request to be formed in the current API. If I were to implement the ability to support this in batch requests, it would require a lot of string manipulation which I'm unsure is suitable for the quality of API, as I would have to generate the multipart string into the request manually. 
Supporting a request that has the `body` variable, while not being multipart is pretty easy to implement though. 
Should I attempt to implement support for multipart requests inside batch requests?

ryanseys commented on lib/batchtransporter.js:
```diff
@@ -0,0 +1,65 @@
+      return opt_callback && opt_callback(new Error('Unexpected response type.'));
+    }
+
+    var boundary_regex = new RegExp('boundary=(.*)$', 'm');
```
Some comments here would be great

ryanseys commented on lib/batchtransporter.js:
```diff
@@ -0,0 +1,65 @@
+        batch_response.push(body);
+      }
+      catch (err) {
+        //do nothing..?
```
Return an error in the callback here if possible. Else do nothing.

ryanseys commented on lib/googleapis.js:
```diff
@@ -65,6 +67,80 @@ GoogleApis.prototype.addAPIs = function(apis) {
+    }
+    body.push('');
+
+
```
Remove second new line

ryanseys commented:
Please add tests to cover this functionality.

ahageali commented:
Implemented support for streams as the request body and also multipart requests.
Tried to test using the upload API but found out (much) later that uploads are not supported through batch requests, so I'm not sure if any API actually exists that allows uploads while still being supported by batch requests.

I've also implemented some tests, though I'm not sure if they cover the functionality properly as I'm actually pretty new to node and haven't gotten into testing yet. Is it alright if the tests make real API calls, as emulating the requests/responses is tough due to the unique nature of the batch requests / responses.

ryanseys commented on package.json:
```diff
@@ -35,6 +35,10 @@
     "gapitoken": "~0.1.2",
+    "http-string-parser": "0.0.4",
+    "memorystream": "^0.3.1",
+    "multipart-stream": "^1.0.0",
```
This module has a bug that fills the entire contents of the stream into memory, we previously used it and now have ripped it out of the library. I don't believe you'll need any of this stream functionality anyway because uploads are not supported by the batch API.

ahageali commented:
I have removed both `multipart-stream` and `memorystream` modules, and gotten rid of any support in the `batch` method for streams.

bentona commented on test/test.batch.js:
```diff
@@ -0,0 +1,42 @@
+      assert(!result);
+      scope.done();
+      done();
+    });
```
How about a test to verify that each request is executed?

blackrabbit99 commented:
@ryanseys  so what is your politics with batches requests? Is it going to add into node js client or no.
I've several thousands requests and it works awful. Or do you recommend to use third party libs or maybe this branch as temporary decision

ryanseys commented:
I no longer make big decisions with regard to this library, sorry. Please direct all future questions to @tbetbetbe. Thank you.

jokesterfr commented:
Hi, any news from this? Any plan to add this kind of feature in the project?
Thank you

blackrabbit99 commented:
Looks like google team decided not support batch activity into their nodeJS library due to some reason.

I found it as a big issue for my project so make some part of batch api.
Take a look https://github.com/blackrabbit99/google-drive-batch-requests

It works with drive api now. It's not so hard to implement batch support for others api types based on http stack.

raz3r commented:
Is there any reason why this library doesn't support batch requests? Some APIs such as Directory API are quite useless without batch support...

JustinBeckwith commented:
Greetings!  And thanks for the contribution.  Sadly - this PR has gotten really stale :(  We're taking steps to respond to these in a more timely fashion to prevent this from happening in the future.  Apologies!

grant commented:
Note: It also looks like global HTTP batch endpoints (www.googleapis.com/batch) will cease to work on March 25, 2019 anyways.

https://developers.googleblog.com/2018/03/discontinuing-support-for-json-rpc-and.html
